sqldelight-gradle-plugin/src/main/kotlin/app/cash/sqldelight/gradle/SqlDelightDatabase.kt

```diff
@@ -41,26 +41,39 @@ abstract class SqlDelightDatabase @Inject constructor(
   abstract val migrationOutputFileFormat: Property<String>
   abstract val generateAsync: Property<Boolean>
 
-  internal val configuration = project.configurations.create("${name}DialectClasspath").apply {
-    isCanBeConsumed = false
-    isVisible = false
+  internal abstract val dialect: Property<Dependency>
+  internal val configuration = project.configurations.register("${name}DialectClasspath") {
+    it.isCanBeConsumed = false
+    it.isVisible = false
+    it.dependencies.addLater(dialect.orElse(project.provider { 
+      error("""
+        A dialect is needed for SQLDelight. For example for sqlite:
+      sqldelight {
+        $name {
+          dialect("app.cash.sqldelight:sqlite-3-18-dialect:$VERSION")
+        }
+      }
+      """.trimIndent())
+    }))
   }
 
-  internal val moduleConfiguration = project.configurations.create("${name}ModuleClasspath").apply {
-    isCanBeConsumed = false
-    isVisible = false
+  internal val moduleConfiguration = project.configurations.register("${name}ModuleClasspath") {
+    it.isCanBeConsumed = false
+    it.isVisible = false
   }
 
-  internal var addedDialect: Boolean = false
-
   fun module(module: Any) {
-    moduleConfiguration.dependencies.add(project.dependencies.create(module))
+    project.dependencies.add(moduleConfiguration.name, module)
   }
 
   fun dialect(dialect: Any) {
-    if (addedDialect) throw IllegalStateException("Can only set a single dialect.")
-    project.dependencies.add(configuration.name, dialect)
-    addedDialect = true
+    this.dialect.set(project.dependencies.create(dialect))
+    this.dialect.finalizeValue()
```

Comment:
or is that what this does?

Reply:
This gives you the error: "The property dialect is already set."